hudi-aws/src/main/java/org/apache/hudi/aws/sync/AWSGlueCatalogClient.java

```diff
@@ -0,0 +1,483 @@
+      schema.putAll(columnsMap);
+      schema.putAll(partitionKeysMap);
+      final long end = System.currentTimeMillis();
+      LOG.info(String.format("Time taken to getTableSchema: %s ms", (end - start)));
+      return schema;
+    } catch (Exception e) {
+      throw new HoodieGlueSyncException("Fail to get schema for table " + tableId(databaseName, tableName), e);
+    }
+  }
+
+  @Override
+  public boolean doesTableExist(String tableName) {
+    return tableExists(tableName);
+  }
+
+  @Override
+  public boolean tableExists(String tableName) {
+    GetTableRequest request = new GetTableRequest()
+        .withDatabaseName(databaseName)
+        .withName(tableName);
+    try {
+      return Objects.nonNull(awsGlue.getTable(request).getTable());
+    } catch (EntityNotFoundException e) {
+      LOG.info("Table not found: " + tableId(databaseName, tableName), e);
+      return false;
+    } catch (Exception e) {
+      throw new HoodieGlueSyncException("Fail to get table: " + tableId(databaseName, tableName), e);
+    }
+  }
+
+  @Override
+  public boolean databaseExists(String databaseName) {
+    GetDatabaseRequest request = new GetDatabaseRequest();
+    request.setName(databaseName);
+    try {
+      return Objects.nonNull(awsGlue.getDatabase(request).getDatabase());
+    } catch (EntityNotFoundException e) {
+      LOG.info("Database not found: " + databaseName, e);
+      return false;
+    } catch (Exception e) {
+      throw new HoodieGlueSyncException("Fail to check if database exists " + databaseName, e);
+    }
+  }
+
+  @Override
+  public void createDatabase(String databaseName) {
+    if (databaseExists(databaseName)) {
+      return;
+    }
+    CreateDatabaseRequest request = new CreateDatabaseRequest();
+    request.setDatabaseInput(new DatabaseInput()
+        .withName(databaseName)
+        .withDescription("automatically created by hudi")
+        .withParameters(null)
+        .withLocationUri(null));
+    try {
+      CreateDatabaseResult result = awsGlue.createDatabase(request);
+      LOG.info("Successfully created database in AWS Glue: " + result.toString());
+    } catch (AlreadyExistsException e) {
+      LOG.warn("AWS Glue Database " + databaseName + " already exists", e);
+    } catch (Exception e) {
+      throw new HoodieGlueSyncException("Fail to create database " + databaseName, e);
+    }
+  }
+
+  @Override
+  public Option<String> getLastCommitTimeSynced(String tableName) {
+    try {
+      Table table = getTable(awsGlue, databaseName, tableName);
+      return Option.of(table.getParameters().getOrDefault(HOODIE_LAST_COMMIT_TIME_SYNC, null));
+    } catch (Exception e) {
+      throw new HoodieGlueSyncException("Fail to get last sync commit time for " + tableId(databaseName, tableName), e);
+    }
+  }
+
+  @Override
+  public void close() {
+    awsGlue.shutdown();
+  }
+
+  @Override
+  public void updateLastCommitTimeSynced(String tableName) {
```

Comment:
A general question about updating HOODIE_LAST_COMMIT_TIME_SYNC. I see in HiveSyncTool, we call hoodieHiveClient.updateLastCommitTimeSynced(tableName) 
in the end. 
But updating partitions and updating the last commit time synced is not atomic. and within updateLastCommitTimeSynced, we get the latest commit time from activetimeline and update it as the lastCommitTimeSycned.  
So, incase of multi-writers, there are chances that timeline has moved between updating partitions and calling updateLastCommitTimeSynced.

may be we should file a jira and do a follow up on the logic.

Reply:
yea, should be a follow up item